src/hevm/src/EVM.hs

```diff
@@ -1880,37 +1883,82 @@ type CheatAction = Word -> Word -> Buffer -> EVM ()
             makeUnique (mksym $ sFromIntegral a) $ \(C _ (num -> a')) ->
               fetchAccount a' $ \_ -> do
                 modifying (env . contracts . ix a' . storage) (writeStorage (mksym slot) (mksym new))
           _ -> vmError (BadCheatCode sig),
 
-      action "load(address,bytes32)" $
-        \sig outOffset _ input -> case decodeStaticArgs input of
+      action "load(address,bytes32)" [AbiAddressType, AbiBytesType 32] $
+        \sig _ outOffset _ input -> case decodeStaticArgs input of
           [a, slot] ->
             makeUnique (mksym $ sFromIntegral a) $ \(C _ (num -> a'))->
               accessStorage a' (mksym slot) $ \res -> do
                 assign (state . returndata . word256At 0) res
                 assign (state . memory . word256At outOffset) res
+          _ -> vmError (BadCheatCode sig),
+
+      action "sign(uint256,bytes)" [AbiUIntType 256, AbiBytesDynamicType] $
+        \sig tps outOffset _ input -> case decodeBuffer tps input of
+          CAbi [AbiUInt 256 sk, AbiBytesDynamic msg] -> let
+              curve = getCurveByName SEC_p256k1
+              priv = PrivateKey curve (num sk)
+              s = ethsign priv msg
+              v = if (sign_s s) % 2 == 0 then 27 else 28
+              encoded = encodeAbiValue $
+                AbiTuple (RegularVector.fromList
+                  [ AbiUInt 8 v
+                  , AbiBytes 32 (encodeAbiValue . AbiUInt 256 . fromInteger $ sign_r s)
+                  , AbiBytes 32 (encodeAbiValue . AbiUInt 256 . fromInteger $ sign_s s)
+                  ])
+            in do
+              assign (state . returndata) (ConcreteBuffer encoded)
+              copyBytesToMemory (ConcreteBuffer encoded) (num . BS.length $ encoded) 0 outOffset
+          _ -> vmError (BadCheatCode sig),
+
+      action "addr(uint256)" [AbiUIntType 256] $
+        \sig _ outOffset _ input -> case decodeStaticArgs input of
+          [sk] -> forceConcrete (mksym sk) $ \sk' -> let
+                curve = getCurveByName SEC_p256k1
+                pubPoint = generateQ curve (num sk')
+                encodeInt = encodeAbiValue . AbiUInt 256 . fromInteger
+              in do
+                case pubPoint of
+                  PointO -> do vmError (BadCheatCode sig)
+                  Point x y -> do
+                    -- See yellow paper #286
+                    let
+                      pub = BS.concat [ encodeInt x, encodeInt y ]
+                      addr = word256 . BS.drop 12 . BS.take 32 . keccakBytes $ pub
+                    assign (state . returndata . word256At 0) (w256lit . num $ addr)
+                    assign (state . memory . word256At outOffset) (w256lit . num $ addr)
           _ -> vmError (BadCheatCode sig)
+
     ]
   where
-    action s f = (abiKeccak s, f (Just $ abiKeccak s))
+    action s tps f = (abiKeccak s, f (Just $ abiKeccak s) tps)
     mksym x = S (Todo "abidecode" []) x
 
+-- | Hack deterministic signing, totally insecure...
```

Comment:
This is pretty sketchy. If we were doing this properly we would either:

1. Implement [rfc-6979](https://tools.ietf.org/html/rfc6979) (perhaps in cryptonite itself).
2. Add a new `Query` that lets us get random numbers

Reply:
I think I'm going to implement #2 next week when we have a bit more free time...

Reply:
Decided to leave it as it is in the end. Signatures produced by `hevm.sign` will leak the private key, but that seems fine for our purposes....